src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdApplicationStep.cs

```diff
@@ -64,29 +80,56 @@ public async Task ExecuteAsync(RecipeExecutionContext context)
-                descriptor.Permissions.Add(OpenIddictConstants.Permissions.Endpoints.Logout);
-            }
             if (model.AllowAuthorizationCodeFlow || model.AllowClientCredentialsFlow ||
                 model.AllowPasswordFlow || model.AllowRefreshTokenFlow)
             {
                 descriptor.Permissions.Add(OpenIddictConstants.Permissions.Endpoints.Token);
             }
 
-            descriptor.PostLogoutRedirectUris.UnionWith(
-                from uri in model.PostLogoutRedirectUris?.Split(' ', StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>()
-                select new Uri(uri, UriKind.Absolute));
+            if (model.AllowAuthorizationCodeFlow)
+            {
+                descriptor.Permissions.Add(OpenIddictConstants.Permissions.ResponseTypes.Code);
+            }
+            if (model.AllowImplicitFlow)
+            {
+                descriptor.Permissions.Add(OpenIddictConstants.Permissions.ResponseTypes.IdToken);
+
+                if (string.Equals(model.Type, OpenIddictConstants.ClientTypes.Public, StringComparison.OrdinalIgnoreCase))
+                {
+                    descriptor.Permissions.Add(OpenIddictConstants.Permissions.ResponseTypes.IdTokenToken);
+                    descriptor.Permissions.Add(OpenIddictConstants.Permissions.ResponseTypes.Token);
+                }
+            }
+            if (model.AllowHybridFlow)
+            {
+                descriptor.Permissions.Add(OpenIddictConstants.Permissions.ResponseTypes.CodeIdToken);
 
-            descriptor.RedirectUris.UnionWith(
-                from uri in model.RedirectUris?.Split(' ', StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>()
-                select new Uri(uri, UriKind.Absolute));
+                if (string.Equals(model.Type, OpenIddictConstants.ClientTypes.Public, StringComparison.OrdinalIgnoreCase))
+                {
+                    descriptor.Permissions.Add(OpenIddictConstants.Permissions.ResponseTypes.CodeIdTokenToken);
+                    descriptor.Permissions.Add(OpenIddictConstants.Permissions.ResponseTypes.CodeToken);
+                }
+            }
 
-            descriptor.Roles.UnionWith(model.RoleEntries
-                .Where(role => role.Selected)
-                .Select(role => role.Name));
+            if (model.AllowAuthorizationCodeFlow || model.AllowHybridFlow || model.AllowImplicitFlow)
+            {
+                if (model.RedirectUris.Any())
+                {
+                    descriptor.RedirectUris.UnionWith(
+                    from uri in model.RedirectUris
+                    select new Uri(uri, UriKind.Absolute));
+                }
+            }
 
-            await _applicationManager.CreateAsync(descriptor);
+            var application = await _applicationManager.FindByClientIdAsync(descriptor.ClientId);
+            if (application != null)
+            {
+                await _applicationManager.UpdateAsync(application, descriptor);
```

Comment:
If you want to support UPSERT, you'll need to make additional changes to this class to remove the permissions/redirects_uris/post_logout_redirect_uris. Take a look at how it's done in the app controller.